cpp/src/Ice/Instance.cpp

```diff
@@ -1283,6 +1301,33 @@ IceInternal::Instance::initialize(const Ice::CommunicatorPtr& communicator)
     }
 }
 
+const Ice::ImplicitContextPtr&
+IceInternal::Instance::getImplicitContext() const
+{
+    switch (_implicitContextKind)
+    {
+        case ImplicitContextKind::PerThread:
+        {
+            static thread_local std::map<const IceInternal::Instance*, ImplicitContextPtr> perThreadImplicitContextMap;
+            auto it = perThreadImplicitContextMap.find(this);
+            if (it == perThreadImplicitContextMap.end())
+            {
+                auto r = perThreadImplicitContextMap.emplace(make_pair(this, std::make_shared<ImplicitContext>()));
+                return r.first->second;
+            }
+            else
+            {
+                return it->second;
+            }
```

Comment:
This code replaces the old `PerThreadImplicitContext` we keep a per thread, per instance ImplicitContext.